src/base/math/gcomplex.h

```diff
@@ -140,10 +140,24 @@ class GPUcomplex {
 
   /// Note: You should not use this operator to compare with zero, because
   /// cmp_rel breaks down in that case.
+  /// ^ perhaps ok now; I added an explicit check for zero ~dsh
   __host__ __device__ bool operator==(const GPUcomplex &op) {
     ////TODO:: THAT PRECISION HAS TO BE CHANGED!!
-    return (cmp_rel<floatT>(this->c.x, op.c.x, 1.e-6, 1.e-6) &&
-            cmp_rel<floatT>(this->c.y, op.c.y, 1.e-6, 1.e-6));
+    bool test1, test2;
+
+    if(this->c.x==0.0 || op.c.x==0.0){
```

Comment:
source for that?  I was mostly using this to debug, although I don't think comparison operators introduce round-off.  I do think it is true that underflows used to be truncated to zero, but this hasn't been true for awhile, and certainly that would happen to the inputs prior to a comparison operator like ==.

Without this line, comparisons to zero did fail silently, so == 0.0 seems better than nothing.  It is true that relative comparison of floats makes no sense for comparisons to zero, even if the norm multiplies the relative difference epsilon.  However, here we have an operator named == acting on floating point numbers, but under the hood it for some reason does not behave how == does for c++ floating point numbers.  I appreciate the desire to write something better than what c++ has for this operator, but it seems confusing from my perspective to ask for == and get back a relative comparison that also drops support for comparison to zero and also fails silently for zero comparison.

In the interest of having zero comparison not fail silently and also not have a=b=0 be the only near zero point supported, though (I mean it should not fail for numbers near zero but succeed for zero), let me offer the following:

we have the current comparison test in cmp_rel as

    if (abs(a-b) / abs(a+b) < rel && abs(a-b) < prec) {

Certainly the main failure mode, without specifying rel and prec, is a,b near zero.  Otherwise, this seems sensible, although without knowing what cmp_rel is supposed to do exactly I can't say for sure if it is correct.

I think the general solution, though, should be something like https://stackoverflow.com/a/32334103 for float comparison.  The current solution we are discussing (with comparison == 0.0) is qualitatively the first graph under "Graphs please?"  I think the approach given in the link is (broadly speaking) what I'd support, and I'd be interested to know if people can think of any improvements (besides generic type support).  This link also shows us how to build cmp_rel, in that cmp_rel should be a function like nearly_equal(a,b), but with an assert that causes cmp_rel to fail on comparisons near zero.

If the above were implemented, the point where relative comparison breaks down becomes machine/implementation/type dependent.  However, use of various types like float already subjects us to this dependence.  If we are near these limits, we should expect a lot of problems, not just in our comparison operator.  This should be an indication that the user is using the wrong floatT, rather than a problem with the function/operator.

Reply:
I just want to mention that also issue #29 exists, which is probably related

Reply:
The current version

if (abs(a-b) / abs(a+b) < rel && abs(a-b) < prec) return true

Will indeed give problems for a+b = 0

The link Dan provided will solve this. 

I think the minimalist way to solve this problem is to change the code into

if ( 2.0*abs(a-b)  < rel*(abs(a)+abs(b))+1e-37 && abs(a-b) < prec) return true

If you don't like the small constant added (chosen to be around 10 times smaller that the smallest float number) one can replace 

rel*(abs(a)+abs(b))+1e-37 with  max(rel*(abs(a)+abs(b)), smallest float number)

Last thing. I think the if statement is not needed, should just be

return (2.0*abs(a-b)  < rel*(abs(a)+abs(b))+1e-37 && abs(a-b) < prec)

This does assume that a and b are smaller than half the maximum float number. A check for this can be added if wanted.